platform/security/idp/security-idp-plugin/security-idp-plugin-nameidpolicy/src/main/java/org/codice/ddf/security/idp/plugin/nameidpolicy/NameIdPolicyPresignPlugin.java

```diff
@@ -0,0 +1,154 @@
+                    response));
+  }
+
+  private void setAssertionNameIdQualifierAndFormat(
+      NameID assertionNameId,
+      @Nullable String nameIdFormatPolicy,
+      @Nullable String spNameQualifierPolicy,
+      Response response) {
+
+    if (StringUtils.isNotEmpty(spNameQualifierPolicy)) {
+      assertionNameId.setSPNameQualifier(spNameQualifierPolicy);
+    }
+
+    // the Assertion's NameID Format is set to the NameIDPolicy Format after the switch statement
+    if (StringUtils.isNotEmpty(nameIdFormatPolicy)) {
+      switch (nameIdFormatPolicy) {
+          // supported NameIDFormats
+        case NameID.UNSPECIFIED:
+          return; // avoid changing the Format later
+        case NameID.PERSISTENT:
+          // TODO DDF-3965
+          break;
+
+          // partially supported NameIDFormats
+        case NameID.X509_SUBJECT:
+          LOGGER.warn("Specifying the \"X509Subject\" NameIDPolicy Format is not fully supported.");
+          if (!NameID.X509_SUBJECT.equals(assertionNameId.getFormat())) {
+            throw new UnsupportedOperationException(
+                "The \"X509Subject\" NameID could not be retrieved.");
+          }
+          break;
+        case NameID.EMAIL:
+          LOGGER.warn("Specifying the \"Email\" NameIDPolicy Format is not fully supported.");
+          assertionNameId.setValue(resolveEmail(response));
+          break;
+
+          // not supported NameIDFormats
+        case NameID.TRANSIENT:
+        case NameID.WIN_DOMAIN_QUALIFIED:
+        case NameID.KERBEROS:
+        case NameID.ENTITY:
+        default:
+          throw new UnsupportedOperationException(
+              String.format(
+                  "The NameIDPolicy Format of value %s is not supported.", nameIdFormatPolicy));
+      }
+      assertionNameId.setFormat(nameIdFormatPolicy);
+    }
+  }
+
+  /**
+   * Attempts to find an email identifier in a given <@code>Response</@code>. More specifically,
+   * this method looks in a <@code>Response</@code>'s <@code>AttributeStatements</@code> for the
+   * email identifier. Throws an error if not found.
+   *
+   * @param response <@code>Response</@code> object
+   * @return Email identifier
+   */
+  private String resolveEmail(Response response) {
+    Predicate<Attribute> attributeHasEmailIdentifier =
+        attribute ->
+            SubjectUtils.EMAIL_ADDRESS_CLAIM_URI.equals(attribute.getName())
+                || NameID.EMAIL.equals(attribute.getNameFormat());
+
+    return extractAssertionAttributes(response)
+        .filter(attributeHasEmailIdentifier)
+        .flatMap(attribute -> attribute.getAttributeValues().stream())
+        .filter(attributeValue -> attributeValue instanceof XSString)
+        .map(attributeValue -> (XSString) attributeValue)
+        .map(XSString::getValue)
+        .filter(StringUtils::isNotBlank)
+        .findFirst()
+        .orElseThrow(
+            () ->
+                new UnsupportedOperationException(
```

Comment:
❓ Should we identify the principal somehow so this can be solved easier?

Reply:
I don't think I understand what you are asking. We already have all of the information about the principal, we just need to see if we have an email identifier in that information.

Reply:
I am suggesting we identify the principal in the exception/log message so that an admin can diagnose the problem easier.